docs: add react sdk blog (#541)
open-feature/openfeature.dev: blog/2024-05-02-react-sdk-release.md

```tsx
{/* suspense boundary and fallback */}
<Suspense fallback={<Fallback />}>
  <Greeting />
</Suspense>

function Fallback() {
  return (<>Loading...</>);
}

function Greeting() {
  // this hook will cause the component to suspend until the provider is ready
  const { value: showEmoji } = useFlag("show-emoji", false);

  return (
    <>
      {showEmoji ? (
        <p>👋 Welcome to this React app!</p>
      ) : (
        <p>Welcome to this React app!</p>
      )}
    </>
  );
}
```

## Compatibility

As previously mentioned, the React SDK is built on top of the existing `@openfeature/web-sdk`.
Any OpenFeature provider built to support the web SDK can also be used with the `@openfeature/react-sdk`, no changes required.

## Future Plans

We have a few enhancements in mind, particularly around mutating evaluation context and an API that provides components (in addition to hooks), but we're excited to get feedback from the community on what we've built so far!
If you're a React dev (or you're just looking for an excuse to familiarize yourself with the library) take the OpenFeature React SDK for a spin and let us know how it goes!

> You can play around with our test app [here](https://github.com/open-feature/react-test-app) to familiarize yourself with some of the features.

Author: I'm in the process of updating the README for this repo.

Author: This is done.

---
slug: "openfeature-react-sdk-release"
title: "OpenFeature React SDK Release"
date: 2024-05-02
authors: [toddbaert]
description: "Announcing the first official release of the OpenFeature React SDK"
tags: [react, client, sdk, web, javascript, typescript]
draft: false
---

> We're stoked to announce the first release of the [@openfeature/react-sdk](https://www.npmjs.com/package/@openfeature/react-sdk)!

Reviewer: I think we should link here instead of NPM.

https://deploy-preview-541--openfeature.netlify.app/docs/reference/technologies/client/web/react

Author: Agreed

This SDK is a distribution of our SDK for the web with extensions to support idiomatic React development and features.
This blog highlights a few of these, with a few implementation details for the real React nerds. 🤓

<!--truncate-->

## Features

### Context Provider

The OpenFeature client is a lightweight abstraction used to evaluate feature flags, implemented by every SDK.
Essentially, they represent a _scope_ for feature flag evaluations.
React's [_context provider_](https://react.dev/reference/react/createContext#provider) concept provides a means of sharing data and objects across multiple components without explicitly passing them through [_props_](https://react.dev/learn/passing-props-to-a-component).

> A context provider is an ideal means of providing a bounded scope for flag evaluations by associating a react context with an OpenFeature client; the OpenFeature React SDKs `OpenFeatureProvider` does exactly this.

Author: ```suggestion
A context provider is an ideal means of providing a bounded scope for flag evaluations by associating a react context with an OpenFeature client; the OpenFeature React SDK's `OpenFeatureProvider` does exactly this.
```

Within the scope of this React provider, all flag evaluation hooks will make use of the bound OpenFeature client and its associated [OpenFeature provider](https://openfeature.dev/specification/glossary#provider).

```tsx
function App() {
  return (
    <OpenFeatureProvider>
      {/* components in this scope share the same OpenFeature client and OpenFeature Provider */}
      <MyPage/>
    </OpenFeatureProvider>
  );
}
```

### Re-rendering with Flag Changes

> The most fundament feature of the React SDK is support for re-rending components using feature flags when the effective values of those flags change.

Reviewer 2: ```suggestion
The most fundamental feature of the React SDK is support for re-rending components using feature flags when the effective values of those flags change.
```

Flag value changes could be the result of changes in the evaluation context which is the basis for flag evaluations rules, or change in the flag definitions themselves.

> Changes in the evaluation context are best exemplified by changes to user properties in a application; for example, a user logging in, adding something to their cart, or changing their subscriber level.

Reviewer: ```suggestion
Changes in the evaluation context are best exemplified by changes to user properties in an application; for example, a user logging in, adding something to their cart, or changing their subscriber level.
```

Any of these examples could mean that we need to re-evaluate feature flags in light of this new information.
The flag evaluation hooks provided by the React SDK do this automatically (by default) when the bound provider emits a [_PROVIDER_CONTEXT_CHANGED_](https://openfeature.dev/specification/types#provider-events) event.

```tsx
function MyPage() {
  // the "value" returned here is always up-to-date with changes in rules or contextual data
  const { value: showNewMessage } = useFlag("new-message", false);
  return (
    <div>
      <header>
        {showNewMessage ? (
          <p>Welcome to this OpenFeature-enabled React app!</p>
        ) : (
          <p>Welcome to this React app.</p>
        )}
      </header>
    </div>
  );
}
```
Additionally, some feature flag solutions support realtime updates when flag configurations themselves are changed and the bound provider emits a [PROVIDER_CONFIGURATION_CHANGED](https://openfeature.dev/specification/types#provider-events) event.

> The React SDK also supports re-rendering in these cases, so that a change in a flag definition results in those changes immediately being reflected in the user interface.

Reviewer: ```suggestion
The React SDK also supports re-rendering in these cases so that a change in a flag definition immediately reflects those changes in the user interface.
```

```tsx
{/* suspense boundary and fallback */}
<Suspense fallback={<Fallback />}>
  <Greeting />
</Suspense>

function Fallback() {
  return (<>Loading...</>);
}

function Greeting() {
  // this hook will cause the component to suspend until the provider is ready
  const { value: showEmoji } = useFlag("show-emoji", false);

  return (
    <>
      {showEmoji ? (
        <p>👋 Welcome to this React app!</p>
      ) : (
        <p>Welcome to this React app!</p>
      )}
    </>
  );
}
```

## Compatibility

As previously mentioned, the React SDK is built on top of the existing `@openfeature/web-sdk`.
Any OpenFeature provider built to support the web SDK can also be used with the `@openfeature/react-sdk`, no changes required.

## Future Plans

We have a few enhancements in mind, particularly around mutating evaluation context and an API that provides components (in addition to hooks), but we're excited to get feedback from the community on what we've built so far!
If you're a React dev (or you're just looking for an excuse to familiarize yourself with the library) take the OpenFeature React SDK for a spin and let us know how it goes!

> You can play around with our test app [here](https://github.com/open-feature/react-test-app) to familiarize yourself with some of the features.

Reviewer: I'll write a tutorial that we can also link to.

Author: Happy to add that when it's ready.
